dipy/align/imaffine.py

```diff
@@ -0,0 +1,715 @@
+        self.prealign = prealign
+        self.param_scales = None
+        static_32 = np.array(static).astype(np.float32)
+        moving_32 = np.array(moving).astype(np.float32)
+
+        T = np.eye(self.dim + 1)
+        if self.prealign is not None:
+            T = T.dot(self.prealign)
+
+        if self.dim == 2:
+            self.interp_method = vf.interpolate_scalar_2d
+        else:
+            self.interp_method = vf.interpolate_scalar_3d
+
+        if self.sampling_proportion is None:
+            self.warped = aff_warp(self.static, self.static_grid2space,
+                                   self.moving, self.moving_grid2space, T)
+            self.warped = self.warped.astype(np.float64)
+            self.samples = None
+            self.ns = 0
+        else:
+            self.warped = None
+            k = 100 / self.sampling_proportion
+            shape = np.array(static.shape, dtype=np.int32)
+            samples = sample_domain_regular(k, shape, static_grid2space)
+            samples = np.array(samples)
+            ns = samples.shape[0]
+            # Add a column of ones (homogeneous coordinates)
+            samples = np.hstack((samples, np.ones(ns)[:, None]))
+            # Sample the static image
+            points_on_static = self.static_space2grid.dot(samples.T).T
+            points_on_static = points_on_static[..., :self.dim]
+            static_vals, inside = self.interp_method(static_32,
+                                                     points_on_static)
+            static_vals = np.array(static_vals, dtype=np.float64)
+            # Sample the moving image
+            sp_to_moving = self.moving_space2grid.dot(T)
+            points_on_moving = sp_to_moving.dot(samples.T).T
+            points_on_moving = points_on_moving[..., :self.dim]
+            moving_vals, inside = self.interp_method(moving_32,
+                                                     points_on_moving)
+            # Store relevant information
+            self.samples = samples
+            self.ns = ns
+            self.static_vals = static_vals
+            self.moving_vals = moving_vals
+
+        MattesBase.setup(self, self.static, self.moving)
+
+    def _update_dense(self, xopt, update_gradient=True):
+        r""" Updates marginal and joint distributions and the joint gradient
+
+        The distributions are updated according to the static and warped
+        images. The warped image is precisely the moving image after
+        transforming it by the transform defined by the xopt parameters.
+
+        The gradient of the joint PDF is computed only if update_gradient
+        is True.
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+        update_gradient : Boolean
```

Comment:
Good to know! I never specified `optional` previously, now I see what PEP8 suggests